src/main/java/com/google/firebase/database/snapshot/BigDecimalNode.java

```diff
@@ -0,0 +1,60 @@
+package com.google.firebase.database.snapshot;
+
+import java.math.BigDecimal;
+
+public class BigDecimalNode extends LeafNode<BigDecimalNode> {
+
+  private final String value;
+
+  public BigDecimalNode(BigDecimal value, Node priority) {
+    this(value == null ? null : value.toString(), priority);
+  }
+
+  BigDecimalNode(String value, Node priority) {
+    super(priority);
+    this.value = value;
+  }
+
+  @Override
+  public Node updatePriority(Node priority) {
+    return new BigDecimalNode(this.value, priority);
+  }
+
+  @Override
+  public Object getValue() {
+    return this.value;
+  }
+
+  @Override
+  public String getHashRepresentation(HashVersion version) {
+    return null;
+  }
+
+  @Override
+  protected LeafType getLeafType() {
+    return LeafType.String;
```

Comment:
I believe this should be `LeafType.Number`

Reply:
And as a heads up, you will have to revise how leaf nodes are compared in the parent `LeafNode.compareTo()` method. You will see that longs and doubles have been special cased there. Perhaps you can implement a new comparison method for all number types, which converts values to `BigDecimal` and then does the comparison.